### PR TITLE
Fix water purification

### LINKED ISSUE #17308: Water purification tabs not working
Running version 5104; for past several days water purification tablets haven't been working.  To test, fill up a bottle with water, and with multiple tabs in inventory, attempt to Activate the tabs from inventory.  "You don't have enough charges in your purifier to purify all the water" is the resulting error message.  Attached is a savegame illustrating this.

[Spiritwood 2016-06-20.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/328933/Spiritwood.2016-06-20.zip)

### PR DESCRIPTION
**Pull request by mugling**
Fixes #17308 

Starts implementing a new family of functions `item::usages_*` which are agnostic about the underlying item.

The long-term plan is for `iuse` and `iuse_actor` to specify number of usages actually performed (typically `0` or `1`, occasionally more) and then the caller of that function to call `item::usages_deplete` which would remove charges or ammo accounting for `ammo_required` per use, UPS etc.

One small balancing change is that water purification tablets nolonger desalinate water as that's an enormously energetic process and entirely non-plausible for chemical sterilization tablets.

### CODE REVIEW

File: src/iuse.cpp [hunk 1]
```diff
@@ -2678,32 +2678,25 @@ static int cauterize_elec(player *p, item *it)
 
 int iuse::water_purifier(player *p, item *it, bool, const tripoint& )
 {
-    auto loc = g->inv_map_splice( []( const item & itm ) {
-        return !itm.contents.empty() &&
-               ( itm.contents.front().typeId() == "water" ||
-                 itm.contents.front().typeId() == "salt_water" );
+    auto obj = g->inv_map_splice( []( const item &e ) {
+        return !e.contents.empty() && e.contents.front().typeId() == "water";
```
Comment (Coolthulhu) [inline src/iuse.cpp]: No more salt water purification?

**Comment by Coolthulhu** [comment]
> The long-term plan is for iuse and iuse_actor to specify number of usages actually performed (typically 0 or 1, occasionally more) and then the caller of that function to call item::usages_deplete which would remove charges or ammo accounting for ammo_required per use, UPS etc.

This sounds like a redundant multiplier. Why have the "usages" when `iuse`s would already specify the cost?

**Comment by mugling** [comment]
@Coolthulhu because 99%+ of iuse functions can just specify `1` or `0` and not worry about how many charges their underlying item uses per invocation (which could for example be influenced by `ammo_required`, UPS or other such

**Comment by Coolthulhu** [comment]
> 99%+ of iuse functions can just specify 1 or 0

But then we're back at square one and still can't have tools with uses with `>0` cost and ones with `0` cost.
The only real way to solve this is per-item, per-iuse cost with no extra multipliers along the way (no shortcuts like "3 times the default cost", except at json loading level, as `"relative"/"proportional"` field).

> The alternative is that logic being moved into each iuse function which is unecessarily complex

Only the multiplication part, which shouldn't really be a thing anyway.

**Comment by mugling** [comment]
> But then we're back at square one and still can't have tools with uses with >0 cost and ones with 0 cost.

Why not? You're confounding usages with cost. I might use my hammer 4 times but it doesn't cost anything?

**Comment by mugling** [comment]
Or a better example a woodsaw with usage 1 has on ammo cost but a circular saw does. The only rational unit to return from an iuse function is how man usages there were - the caller is responsible for consuming any ammo/charges and/or destroying the item

File: src/item.cpp [hunk 2]
```diff
@@ -4393,6 +4393,27 @@ item *item::get_usable_item( const std::string &use_name )
...
 }
 
+long item::usages_remaining( const Character& ch ) const
+{
+    // items counted by charge always consume one charge per usage
+    if( count_by_charges() ) {
+        return charges;
+    }
+
+    // for items not consuming ammo then allow infinite usages
```
Comment (BevapDin) [inline src/item.cpp]: Things like traps get used up instead. They consume no ammo, so `ammo_required()` should return 0.

File: src/item.cpp [hunk 3]
```diff
@@ -4393,6 +4393,27 @@ item *item::get_usable_item( const std::string &use_name )
...
+    auto req = ammo_required();
+    if( req == 0 ) {
+        return LONG_MAX;
+    }
+
+    // for tools modified to use UPS instead limit by available UPS charges
```
Comment (BevapDin) [inline src/item.cpp]: The charges contained in the item itself also count. One can use a charged UPS compatible tool without having a UPS in the inventory.

File: src/item.cpp (hunk 2 above)
Comment (mugling) [inline src/item.cpp]: How do you detect than item is a trap?

**Comment by Coolthulhu** [comment]
> The only rational unit to return from an iuse function is how man usages there were - the caller is responsible for consuming any ammo/charges and/or destroying the item

Not at all. "Usage" is a very broad term that only really works when all uses have identical costs.
With "usages", the only rational way to implement everything is to drop `ammo_required` to 1 and return as many usages as the cost should be, thus working around the clunky multiplication by `ammo_required`.

It sounds like you're trying to hack in `charges_to_use` into some bigger system. But the proper solution involves getting rid of `charges_to_use` in its current form and moving it to JSON.
In case of tools, `ammo_required` is just `charges_to_use`.

For it to work, `ammo_required` would have to take an `iuse` parameter, to get per-iuse charges from it. For as long as `ammo_required` is only dependent on the item itself, nothing really changes here, except for additional layer of abstraction to take into account.

File: src/item.cpp (hunk 2 above)
Comment (BevapDin) [inline src/item.cpp]: It's handled specially in `player::consume_charges`. The iuse function returns 1 or something like that as usual.

**Comment by mugling** [comment]
For some example items using `iuse::water_purifier`:

| item                             | type                | ammo_required() | flags         |
|-----------------------------|-------------------|-----------------------|--------------|
| water purifier               | TOOL             | 1                          |                 |
| makeshift purifier         | TOOL             | 2                           |                 |
| makeshift UPS purifier | TOOL             | 2                           | USE_UPS |
| purification tablet         | COMESTIBLE | 0                           |                 |
| radiation source*         | TOOL             | 0                           |                  |

`*` An arbitrary tool that uses no ammo

I'm proposing that `iuse::water_purifier` returns `usages` - in this case the quantity of `water` acted upon. I think you're proposing that it returns the number of charges consumed by the base item.

So for 4 charges of water the return values from `iuse::water_purifier` for each system would be:

| item                             | mugling               | coolthulhu         |
|-----------------------------|-------------------|-----------------------|
| water purifier               | 4                    | 4                          |
| makeshift purifier         | 4                     | 8                         |
| makeshift UPS purifier | 4                     | 8?                           |
| purification tablet         | 4                    | 4                           |
| radiation source          | 4                     | 0                          |

I can obtain all of those values by a single check against `usages_remaining` followed by returning liquid charges. You would need to implement the entire logic for tools, comestible and UPS within the `iuse function`.

In the caller I can then pass that usage count directly to `item::deplete_usages` whereas you need again to switch between handling the different variants to know what to do with that number?

**Comment by Coolthulhu** [comment]
> I'm proposing that iuse::water_purifier returns usages - in this case the quantity of water acted upon. I think you're proposing that it returns the number of charges consumed by the base item.

You're adding another layer of abstraction here. Why handle cost multiplication as a separate hardcoded function instead of just allowing it naturally as a part of the iuse?
This is restricting and sounds like a feature to work around rather than with.

Instead, that multiplication can be pushed into the `iuse` structure. This could be more work at first, but would end up a less restrictive system that would require less maintenance and be less prone to bugs.

**Comment by mugling** [comment]
That's a **lot** of code duplication in each `iuse` function - at a minimum each would need to:

1. is item counted by charges?
2. does item use ammo at all?
3. is item powered by UPS?
4. does player have UPS?
5. if UPS how many charges are in the item already?
6. is sufficient ammo available?
7. calculate the number of charges and return that.

The caller then needs to repeat the above to remove the charges.

Any changes to any of the above mechanisms would require every `iuse` and `iuse_actor` to be updated. That would be exceptionally painful.

There isn't a good reason why the `iuse` function should care how an item is powered - only that sufficient usages remain. The water purifier only needs to know that the item is capable of being used 4 times (to purify 4 charges) and return how many times it was used.

This is much simpler and doesn't preclude anything as far as I can see. It also reduces almost all `iuse` functions to a boolean `1` or `0` return.

**Comment by Coolthulhu** [comment]
> is item counted by charges?

That's already handled above it. Or was, before. No need for an extra layer of abstraction.
> does item use ammo at all?
> is item powered by UPS?
> does player have UPS?
> if UPS how many charges are in the item already?

Handled just fine as it is.
> is sufficient ammo available?

This must not be a function of the item alone. It must depend on the iuse too, meaning that extracting out an extra layer will not help here.

> calculate the number of charges and return that.

The only thing that changes here is that the multiplication is done in the iuse or the factory that emits it, meaning it is easier to understand and less prone to bugs.

> The caller then needs to repeat the above to remove the charges.

That is already done, you're only adding a multiplier here - one that isn't needed since it doesn't remove much complexity, but is limiting.

> Any changes to any of the above mechanisms would require every iuse and iuse_actor to be updated.

This would be the case with the `usages` system, not with the per-item `iuse` dictating full cost system.

> There isn't a good reason why the iuse function should care how an item is powered - only that sufficient usages remain.

However the iuse must know that sufficient usages **of this particular iuse** are available. Not of some arbitrary unit of usage.

> The water purifier only needs to know that the item is capable of being used 4 times (to purify 4 charges) and return how many times it was used.

An oxytorch needs to know it has 10 uses of "cut steel", 40 uses of "repair metal vehicle", 5 uses of "forge metal item" etc.

> It also reduces almost all iuse functions to a boolean 1 or 0 return.

And makes it harder and hackier to actually break this mold and have items with diverse, realistic costs.

**Comment by mugling** [comment]
> An oxytorch needs to know it has 10 uses of "cut steel", 40 uses of "repair metal vehicle", 5 uses of "forge metal item" etc.

Those are two separate iuse functions and a crafting recipe.

**Comment by mugling** [comment]
>  is item powered by UPS?
> > Handled just fine as it is.

Lots of outstanding bugs.

Latest commit changes `usages_*` to `units_*` and returns raw units (sans multiplier). The convenience function `units_sufficient` tries very hard to avoid subtle bugs and ensure efficiency. For example a UPS powered water purifier should work correctly.

`iuse` functions can now specify `units_sufficient( 10 )` if they have an exact charges cost in mind or `units_sufficient()` to use the base item defaults

**Comment by Coolthulhu** [comment]
Conflicts (non-trivially) with #17627

I'll test #17627 first.

**Comment by mugling** [comment]
Ok, let me know

**Comment by mugling** [comment]
Rebased

**Comment by Coolthulhu** [comment]
* Spawn contained water
* Spawn water purifier tablets
* Drop all but one tablet
* Use tablet on water
* Works despite being one charge of tablet for 2 charges of water

File: src/iuse.cpp [hunk 4]
```diff
@@ -2678,32 +2678,25 @@ static int cauterize_elec(player *p, item *it)
...
-        p->add_msg_if_player(m_info, _("You can only purify water."));
-        return 0;
-    }
 
-    item *pure = &target->contents.front();
-    if( pure->charges > it->ammo_remaining() ) {
-        p->add_msg_if_player(m_info,
-                             _("You don't have enough charges in your purifier to purify all of the water."));
+    item &liquid = obj->contents.front();
+    if( it->units_sufficient( *p, liquid.charges ) ) {
```
Comment (Coolthulhu) [inline src/iuse.cpp]: Looks like it's inverted here

File: src/item.cpp [hunk 5]
```diff
@@ -4401,6 +4401,29 @@ item *item::get_usable_item( const std::string &use_name )
...
+    return res;
+}
+
+bool item::units_sufficient( const Character &ch, int qty ) const
+{
+    if( qty < 0 ) {
+        qty = count_by_charges() ? 1 : ammo_required();
+    }
+
+    return units_remaining( ch, qty ) == qty;
```
Comment (Coolthulhu) [inline src/item.cpp]: This won't return true for `count_by_charges` items with `charges > qty`.

**Comment by mugling** [comment]
Fixed both

**Pull request merged**